contrib/babelfishpg_tsql/src/pl_exec-2.c

```diff
@@ -3661,101 +3662,148 @@ get_insert_bulk_kilobytes_per_batch()
 	return insert_bulk_kilobytes_per_batch;
 }
 
-static int
-exec_stmt_grantschema(PLtsql_execstate *estate, PLtsql_stmt_grantschema *stmt)
+void exec_gen_grantschema(char *schema_name, char *rolname,  PLtsql_stmt_grantschema *stmt, char *priv_name)
 {
 	List	   *parsetree_list;
 	ListCell   *parsetree_item;
+	parsetree_list = gen_grantschema_subcmds(schema_name, rolname, stmt->is_grant, stmt->with_grant_option, priv_name);
+				/* Run all subcommands */
+				foreach(parsetree_item, parsetree_list)
+				{
+					Node	   *stmt = ((RawStmt *) lfirst(parsetree_item))->stmt;
+					PlannedStmt *wrapper;
+
+					/* need to make a wrapper PlannedStmt */
+					wrapper = makeNode(PlannedStmt);
+					wrapper->commandType = CMD_UTILITY;
+					wrapper->canSetTag = false;
+					wrapper->utilityStmt = stmt;
+					wrapper->stmt_location = 0;
+					wrapper->stmt_len = 0;
+
+					/* do this step */
+					ProcessUtility(wrapper,
+								"(GRANT SCHEMA )",
+								false,
+								PROCESS_UTILITY_SUBCOMMAND,
+								NULL,
+								NULL,
+								None_Receiver,
+								NULL);
+
+					/* make sure later steps can see the object created here */
+					CommandCounterIncrement();
+				}
+}
+
+static int
+exec_stmt_grantschema(PLtsql_execstate *estate, PLtsql_stmt_grantschema *stmt)
+{
 	char	   *dbname = get_cur_db_name();
 	char	   *login = GetUserNameFromId(GetSessionUserId(), false);
 	bool		login_is_db_owner;
 	Oid			datdba;
 	char		*rolname;
 	char		*schema_name;
 	ListCell   *lc;
-	ListCell	*lc1;
+	int16 privilege_maskInt = 0;
 	Oid 		schemaOid;
 
+	/* create integer bitmask for permissions*/
+	foreach(lc, stmt->privileges)
+	{
+		char	*priv_name = (char *) lfirst(lc);
+		if (!strcmp(priv_name,"execute"))
+			privilege_maskInt |= 32;
+		if (!strcmp(priv_name,"select"))
+			privilege_maskInt |= 16;
+		if (!strcmp(priv_name,"insert"))
+			privilege_maskInt |= 8;
+		if (!strcmp(priv_name,"update"))
+			privilege_maskInt |= 4;
+		if (!strcmp(priv_name,"delete"))
+			privilege_maskInt |= 2;
+		if (!strcmp(priv_name,"references"))
+			privilege_maskInt |= 1;
+	}
+
```

Comment:
For every bit, there is an equivalent bit value like for execute privilege 32 integer value i..e 100000 in boolean is assigned, for select privilege 16 integer value i..e 10000 in boolean is assigned, for insert privilege 8 integer value i..e 1000 in boolean is assigned, for update privilege 4 integer value i..e 100 in boolean is assigned, for delete privilege 2 integer value i..e 10 in boolean is assigned, for references privilege 1 integer value i..e 1 in boolean is assigned,